### PR TITLE
fix(client): Fix a bug where it causes Node process to crash when calling `emit()`, `disconnect()` and `connect()` in quick succession

### PR DESCRIPTION
**Pull request by raytung**
Signed-off-by: Ray Tung <ray@kasada.io>

### CODE REVIEW

**Comment by treykasada** [comment]
Shouldn't this PR be targeting the upstream repo? Or am I misunderstanding? 😮

**Comment by raytung** [comment]
@treykasada My thinking is to merge this into main so we can reference this in our projects, then simultaneously target this to the upstream repo.

**Comment by raytung** [comment]
@treykasada My thinking is to merge this into master and then raise a PR targeting upstream repo. What do you think? Or should we batch it up @rganatra-kasada's fix as well in this PR and target upstream?

**Comment by rganatra-kasada** [comment]
> @treykasada My thinking is to merge this into master and then raise a PR targeting upstream repo. What do you think? Or should we batch it up @rganatra-kasada's fix as well in this PR and target upstream?

Makes sense to merge this and raise a PR upstream. I think my PR fix might be better off as its own PR, just for the sake of simplicity.

**Review by rganatra-kasada — APPROVED** [review]

**Comment by treykasada** [comment]
@raytung Hmm... Well, the typical process for contributing to projects is to create a branch on your fork, raise a PR from your fork's branch to the upstream repo's `main` branch, and then optionally pull the latest `main` from upstream into your fork once the PR is merged. So unless we're planning on deviating from that process, the only reasons I can imagine we'd want to merge to this fork's `main` branch are:
- We want to use and maintain this fork separately from the upstream project
- We want to batch up multiple PRs worth of changes into a single PR to the upstream repo

I'd...honestly rather we not do either of those things if possible. 😅  The first means we have another project to maintain, and the second reduces the chances of our contributions getting reviewed and merged.

TL;DR: Unless we're planning to separately maintain this fork, I don't see a reason for this PR _not_ to target the upstream repo (although it'll probably be a little weird given you've already opened a PR from your personal fork).

**Comment by rganatra-kasada** [comment]
> @raytung Hmm... Well, the typical process for contributing to projects is to create a branch on your fork, raise a PR from your fork's branch to the upstream repo's `main` branch, and then optionally pull the latest `main` from upstream into your fork once the PR is merged. So unless we're planning on deviating from that process, the only reasons I can imagine we'd want to merge to this fork's `main` branch are:
> 
>     * We want to use and maintain this fork separately from the upstream project
> 
>     * We want to batch up multiple PRs worth of changes into a single PR to the upstream repo
> 
> 
> I'd...honestly rather we not do either of those things if possible. sweat_smile The first means we have another project to maintain, and the second reduces the chances of our contributions getting reviewed and merged.
> 
> TL;DR: Unless we're planning to separately maintain this fork, I don't see a reason for this PR _not_ to target the upstream repo (although it'll probably be a little weird given you've already opened a PR from your personal fork).

Fair points. Perhaps we can maintain a separate branch on this repo with all our fixes for us to reference in our project, but also keep individual PRs against the original repo for merging?

**Comment by raytung** [comment]
Upstream has merged in our changes https://github.com/fluent/fluent-logger-forward-node/pull/34
